luigi/contrib/s3.py

```diff
@@ -163,236 +181,157 @@ def remove(self, path, recursive=True):
-        s3_key.key = key
-        contents = s3_key.get_contents_as_string()
-
-        return contents
-
-    def copy(self, source_path, destination_path, threads=100, start_time=None, end_time=None, part_size=67108864, **kwargs):
+    def copy(self, source_path, destination_path, threads=100, start_time=None, end_time=None, part_size=8388608, **kwargs):
         """
         Copy object(s) from one S3 location to another. Works for individual keys or entire directories.
-
         When files are larger than `part_size`, multipart uploading will be used.
-
         :param source_path: The `s3://` path of the directory or key to copy from
         :param destination_path: The `s3://` path of the directory or key to copy to
         :param threads: Optional argument to define the number of threads to use when copying (min: 3 threads)
         :param start_time: Optional argument to copy files with modified dates after start_time
         :param end_time: Optional argument to copy files with modified dates before end_time
-        :param part_size: Part size in bytes. Default: 67108864 (64MB), must be >= 5MB and <= 5 GB.
-        :param kwargs: Keyword arguments are passed to the boto function `copy_key`
-
+        :param part_size: Part size in bytes
+        :param kwargs: Keyword arguments are passed to the boto function `copy` as ExtraArgs
         :returns tuple (number_of_files_copied, total_size_copied_in_bytes)
         """
+
         start = datetime.datetime.now()
 
         (src_bucket, src_key) = self._path_to_bucket_and_key(source_path)
         (dst_bucket, dst_key) = self._path_to_bucket_and_key(destination_path)
 
-        # As the S3 copy command is completely server side, there is no issue with issuing a lot of threads
-        # to issue a single API call per copy, however, this may in theory cause issues on systems with low ulimits for
-        # number of threads when copying really large files (e.g. with a ~100GB file this will open ~1500
-        # threads), or large directories. Around 100 threads seems to work well.
+        # don't allow threads to be less than 3
+        threads = 3 if threads < 3 else threads
+        import boto3
 
-        threads = 3 if threads < 3 else threads  # don't allow threads to be less than 3
+        transfer_config = boto3.s3.transfer.TransferConfig(
+            max_concurrency=threads, multipart_chunksize=part_size)
         total_keys = 0
 
-        copy_pool = ThreadPool(processes=threads)
-
         if self.isdir(source_path):
-            # The management pool is to ensure that there's no deadlock between the s3 copying threads, and the
-            # multipart_copy threads that monitors each group of s3 copy threads and returns a success once the entire file
-            # is copied. Without this, we could potentially fill up the pool with threads waiting to check if the s3 copies
-            # have completed, leaving no available threads to actually perform any copying.
```

Comment:
Does boto3's `copy` remove this concern?

Reply:
It is indeed. Since boto3 now handles transfers and threads for us, the idea is to delegate the work to it. I have kept the interface the same (users can choose number of threads if they want to) by creating the appropriate TransferConfig object (which is used by boto3's copy)

http://boto3.readthedocs.io/en/latest/reference/customizations/s3.html#boto3.s3.transfer.TransferConfig 

http://boto3.readthedocs.io/en/latest/reference/services/s3.html#S3.Client.copy